src/Compilers/Core/Portable/SourceGeneration/SyntaxStore.cs

```diff
@@ -0,0 +1,130 @@
+            {
+                _compilation = compilation;
+                _syntaxInputNodes = syntaxInputNodes;
+                _enableTracking = enableTracking;
+                _previous = previousStore;
+                _cancellationToken = cancellationToken;
+            }
+
+            public IStateTable GetSyntaxInputTable(SyntaxInputNode syntaxInputNode, NodeStateTable<SyntaxTree> syntaxTreeTable)
+            {
+                Debug.Assert(_syntaxInputNodes.Contains(syntaxInputNode));
+
+                // when we don't have a value for this node, we update all the syntax inputs at once
+                if (!_tableBuilder.Contains(syntaxInputNode))
+                {
+                    // CONSIDER: when the compilation is the same as previous, the syntax trees must also be the same.
+                    // if we have a previous state table for a node, we can just short circuit knowing that it is up to date
+                    // This step isn't part of the tree, so we can skip recording.
+                    var compilationIsCached = _compilation == _previous._compilation;
+
+                    // get a builder for each input node
+                    var syntaxInputBuilders = ArrayBuilder<(SyntaxInputNode node, ISyntaxInputBuilder builder)>.GetInstance(_syntaxInputNodes.Length);
+                    foreach (var node in _syntaxInputNodes)
+                    {
+                        // TODO: We don't cache the tracked incremental steps in a manner that we can easily rehydrate between runs,
+                        // so we'll disable the cached compilation perf optimization when incremental step tracking is enabled.
+                        if (compilationIsCached && !_enableTracking && _previous._tables.TryGetValue(node, out var previousStateTable))
+                        {
+                            _tableBuilder.SetTable(node, previousStateTable);
+                        }
+                        else
+                        {
+                            syntaxInputBuilders.Add((node, node.GetBuilder(_previous._tables, _enableTracking)));
+                        }
+                    }
+
+                    if (syntaxInputBuilders.Count > 0)
+                    {
+                        GeneratorRunStateTable.Builder temporaryRunStateBuilder = new GeneratorRunStateTable.Builder(_enableTracking);
+
+                        // at this point we need to grab the syntax trees from the new compilation, and optionally diff them against the old ones
+                        NodeStateTable<SyntaxTree> syntaxTreeState = syntaxTreeTable;
+
+                        // update each tree for the builders, sharing the semantic model
+                        foreach ((var tree, var state, var syntaxTreeIndex, var stepInfo) in syntaxTreeState)
+                        {
+                            var root = new Lazy<SyntaxNode>(() => tree.GetRoot(_cancellationToken));
```

Comment:
This isn't related to your changes, but it seems like it would be simpler (fewer allocations) if we simply passed the 'tree' instead of 'root' to the builder here and have the builder call GetRoot. If a syntax tree is doing a parse in response to a GetRoot call it is usually caching the root afterwards anyway. Right?

Reply:
It *probably* would, but its hard to be sure. In the case of a recoverable syntax tree the root is held via WeakReference, so it might stick around for the next generator, but under high memory pressure it also might get swapped out before the next walker gets to it. We should probably profile this before making any changes to it.